bin/cron/hoc_student_name_cleanup

```diff
@@ -0,0 +1,224 @@
+  end
+
+  opts.on('--end-date=2021-12-31',
+    Date,
+    'End of interval to clean up (exclusive). If not specified, the script will ' \
+    'calculate this value from the retention period.'
+  ) do |end_date|
+    options[:end_date] = end_date
+  end
+
+  opts.on('--batch_size=100',
+    Integer,
+    'Number of rows in each batch. Defaults to 100 if not specified.'
+  ) do |batch_size|
+    options[:batch_size] = batch_size
+  end
+
+  opts.on('--sleep=0',
+    Integer,
+    'Amount of time in milliseconds to sleep between batches. Higher values ' \
+    'will reduce the performance impact on the datbase. Defaults to 0 msecs ' \
+    'if not specified.'
+  ) do |sleep_msecs|
+    options[:sleep_msecs] = sleep_msecs
+  end
+
+  opts.on('--dry-run',
+    'Enables read-only mode where no changes are written to the database'
+  ) do |dry_run|
+    options[:dry_run] = dry_run
+  end
+
+  opts.on('--log-sql',
+    'Logs all sql statements sent to the database'
+  ) do |log_sql|
+    options[:log_sql] = log_sql
+  end
+
+  opts.on('--[no-]slack',
+    'Sends summary of script to slack. Defaults to true.'
+  ) do |notify_slack|
+    options[:notify_slack] = notify_slack
+  end
+
+  opts.on('-h', '--help', 'Prints this help message') do
+    puts opts
+    exit
+  end
+end.parse!
+CDO.log.info "Initial parsed options: #{options}"
+
+PEGASUS_DB.loggers << CDO.log if options[:log_sql]
+
+if options[:start_date].nil?
+  last_processed_date = Properties.get(LAST_PROCESSED_DATE_PROPERTY)
+  if last_processed_date.nil?
+    CDO.log.error "--start-date is required on first run of the script"
+    return
+  end
+  options[:start_date] = Date.parse(last_processed_date) + 1
+end
+
+if options[:end_date].nil?
+  options[:end_date] = Date.today - options[:retention_days]
+end
+
+# Options should not change from here on
+options.freeze
+CDO.log.info "Final calculated options: #{options}"
+
+# Main loop - Conceptually, we want to remove names from all rows with a
+# finished_at time between the start date and end date. We iterate over the
+# rows with a set of nested loops. The outer loop processes one day at a time,
+# checkpointing after each iteration by updating LAST_PROCESSED_DATE_PROPERTY
+# after each iteration. The inner loop processes the rows for a given day in
+# small batches to minimize impact to the database.
+date = options[:start_date]
+total_rows_processed = 0
+total_rows_updated = 0
+while date < options[:end_date]
+  CDO.log.info "Processing rows with finished_at=#{date}"
+
+  batch_start_time = date.to_time
+  batch_start_id = 0
+  batch_end_time = (date + 1).to_time
+  batch_number = 0
+  batch_rows_updated = 0
+
+  while batch_start_time < batch_end_time
+    ids_to_cleanup = []
+    last_row = nil
+
+    # Construct the sql where clauses for the start and end of the interval for
+    # this batch. Rows are ordered by finished_at, then id so intervals are
+    # described by (finished_at, id) tuples. The id for the end of the interval
```

Comment:
can you explain this sentence and how it relates to the code please? `The id for the end of the interval is always 0.`

Reply:
Oh, good question.  The where clause for the start of the interval looks like this:
`(finished_at = '2020-08-04 23:54:05' AND id >= 999877376) OR finished_at > '2020-08-04 23:54:05')`
It has two parts OR'd together because I sometimes needs to start the interval at a non-zero id.

If I needed to support an non-zero end id, the where clause for the end of the interval would be something like:
`(finished_at < '2020-08-05 00:00:00' OR (DATE(finished_at) = '2020-08-05' AND id < 123456789))`
(ignoring for the moment that DATE() isn't a real sql function).

But since the end id is always zero, I can simplify the where clause for the end of the interval to just be:
`(finished_at < '2020-08-05 00:00:00')`
because no row will have id < 0.

Does that help clarify or still confusing?

Reply:
I attempted to clarify this a bit in the comment (but it's a bit hard to explain this succinctly).

Reply:
i have to admit i still don't understand. why is the end id zero? wouldn't it by definition be larger than the start id?

Reply:
The rows are ordered first by finished_at, then by id (so id only comes into play if two rows have the same finished_at time).